src/Hub.cpp

```diff
@@ -289,7 +315,49 @@ const sc2::Point3D* Hub::GetNextExpansion() {
         return nullptr;
 
     it->owner = Owner::CONTESTED;
-    return &(it->town_hall_location);
+    return &(*it);
+}
+
+Expansion* Hub::GetExpansionOfTownhall(const sc2::Unit& unit_) {
+    for (auto& i : m_expansions) {
+        if (std::floor(i.town_hall_location.x) == std::floor(unit_.pos.x) &&
+            std::floor(i.town_hall_location.y) == std::floor(unit_.pos.y)) {
+            return &i;
+        }
+    }
+
+    return nullptr;
+}
+
+void Hub::CapturedExpansion(const sc2::Unit& unit_) {
+    auto expansion = GetExpansionOfTownhall(unit_);
+    if (!expansion)
+        return;
+
+    // NOTE (impulsecloud): should be transitioning from CONTESTED.
+    // worker_tag should already be set, and still valid
+    expansion->owner = Owner::SELF;
+    expansion->town_hall_tag = unit_.tag;
+}
+
+void Hub::EnemyOwnsExpansion(const sc2::Unit& unit_) {
```

Comment:
Looks like a 'SetOwner' method which should be part of Expansion class.

Reply:
Created SetOwner/RemoveOwner in Expansion.

Reply:
Are CapturedExpansion/EnemyOwnsExpansion okay as separate functions? (I think in the future, the EnemyOwnsExpansion may be a useful place to put reactionary code)

Otherwise, I can merge them to Hub::SetExpansionOwner(unit) then check alliance to decide which owner to pass to Expansion::SetOwner(unit_, owner_). I didn't know how un-generic you wanted it.